<!DOCTYPE html>
<html lang="en-us">
  <head>
    <meta charset="UTF-8">
    <title>Illumina-array-protocols by KHP-Informatics</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" type="text/css" href="stylesheets/normalize.css" media="screen">
    <link href='http://fonts.googleapis.com/css?family=Open+Sans:400,700' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/github-light.css" media="screen">
  </head>
  <body>
    <section class="page-header">
      <h1 class="project-name">Illumina-array-protocols</h1>
      <h2 class="project-tagline">processing illumina SNP arrays</h2>
      <a href="https://github.com/KHP-Informatics/illumina-array-protocols" class="btn">View on GitHub</a>
      <a href="https://github.com/KHP-Informatics/illumina-array-protocols/zipball/master" class="btn">Download .zip</a>
      <a href="https://github.com/KHP-Informatics/illumina-array-protocols/tarball/master" class="btn">Download .tar.gz</a>
    </section>

    <section class="main-content">
      <h1>
<a id="illumina-array-protocols" class="anchor" href="#illumina-array-protocols" aria-hidden="true"><span class="octicon octicon-link"></span></a>
<a href="http://www.illumina.com/">illumina</a>-array-protocols</h1>

<p><strong>Protocols/scripts for processing illumina SNP arrays</strong>  </p>

<p><strong>VERSION: v0.1</strong>  </p>

<p>A set of scripts and protocols that we use to processing raw Illumina SNP array data.</p>

<ul>
<li>BWA Mapping of probe sequences<br>
</li>
<li>Genomestudio SOP (Manual Calling &amp; QC)<br>
</li>
<li>Standard QC (PLINK, bash...) and re-calling No-Calls using zCall<br>
</li>
</ul>

<h2>
<a id="the-team" class="anchor" href="#the-team" aria-hidden="true"><span class="octicon octicon-link"></span></a>The Team</h2>

<p><strong>Bioinformatics</strong> - Hamel Patel, Amos Folarin &amp; Stephen Newhouse  @ <a href="">bluecell.io</a><br>
<strong>Lab</strong> - Charles Curtis &amp; Team @ <a href="http://www.kcl.ac.uk/ioppn/depts/mrc/research/The-IoPPN-Genomics--Biomarker-Core-Facility.aspx">The IoPPN Genomics &amp; Biomarker Core Facility</a>  </p>

<p><img src="./figs/CSProLogo-new-Cropped-313x105.png" alt="illuminaCSPro"></p>

<h2>
<a id="illumina-downloads-resource" class="anchor" href="#illumina-downloads-resource" aria-hidden="true"><span class="octicon octicon-link"></span></a>Illumina Downloads Resource</h2>

<p>This links takes you to Illumina's download page, which provides access to product documentation and
manifests.</p>

<ul>
<li><a href="http://support.illumina.com/downloads.html">Illumina Downloads</a></li>
</ul>

<h2>
<a id="beadchips" class="anchor" href="#beadchips" aria-hidden="true"><span class="octicon octicon-link"></span></a>BeadChips</h2>

<p>These are the BeadChips we have experience in processing so far....</p>

<table>
<thead>
<tr>
<th>Array</th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="http://support.illumina.com/downloads/humanomniexpress-24-v1-0-product-files.html">HumanOmniExpress-24 v1.0 BeadChip</a></td>
</tr>
<tr>
<td><a href="http://support.illumina.com/downloads/humanomniexpress-24-v1-0-product-files.html">HumanOmniExpress-24 v1.1 BeadChip</a></td>
</tr>
<tr>
<td><a href="http://support.illumina.com/downloads/humancoreexome-24-v1-0-product-files.html">HumanCoreExome-24 v1.0 BeadChip</a></td>
</tr>
<tr>
<td><a href="http://support.illumina.com/downloads/humanomniexpressexome-8v1-1_product_files.html">HumanOmniExpressExome-8 v1.1 BeadChip</a></td>
</tr>
<tr>
<td>
<a href="link">MEGA_Consortium</a> (Early Access...)</td>
</tr>
</tbody>
</table>

<h2>
<a id="1-bwa-mapping-of-probe-sequences" class="anchor" href="#1-bwa-mapping-of-probe-sequences" aria-hidden="true"><span class="octicon octicon-link"></span></a>1. BWA Mapping of Probe Sequences</h2>

<p>Illumina SNP arrays include a lot of probes that map to multiple (&gt;500) sites in the Genome.  </p>

<p>For each array we map the probe sequences to the relevant genome build using BWA (as indicated by Illumina manifests), and
identify probes that map 100% to multiple regions (&gt;1 hit) of the genome.</p>

<p>These probes are either flagged for removal before re-calling, or depending on what the data looks likes in Genomestudio,
are zeroed at the Genomestudio stage before clustering.  </p>

<p>Those familiar with processing Illumina Arrays, will see that a lot of the probes we identify are consitently poorly clustered variants, variants not called for alot of samples, and those messy clouds and/or the always homozygous variants - no matter the population or number of samples. </p>

<p>More details soon....  </p>

<h3>
<a id="illumina-array-annotations" class="anchor" href="#illumina-array-annotations" aria-hidden="true"><span class="octicon octicon-link"></span></a>Illumina Array Annotations</h3>

<table>
<thead>
<tr>
<th>Array</th>
<th>Download Link</th>
</tr>
</thead>
<tbody>
<tr>
<td>MEGA_Consortium_15063755_B2.csv</td>
<td><a href="https://s3-eu-west-1.amazonaws.com/illumina-probe-mappings/mega_array_annotations.txt.gz">download</a></td>
</tr>
<tr>
<td>HumanCoreExome-24v1-0_A.csv</td>
<td>download</td>
</tr>
<tr>
<td>HumanOmniExpressExome-8-v1-1-C.csv</td>
<td>download</td>
</tr>
<tr>
<td>XXX</td>
<td><a href="">download</a></td>
</tr>
<tr>
<td>XXX</td>
<td><a href="">download</a></td>
</tr>
<tr>
<td>XXX</td>
<td><a href="">download</a></td>
</tr>
</tbody>
</table>

<h3>
<a id="bwa-mapping-results" class="anchor" href="#bwa-mapping-results" aria-hidden="true"><span class="octicon octicon-link"></span></a>BWA Mapping Results</h3>

<p><strong>Running BWA</strong><br>
Using <a href="https://github.com/KHP-Informatics/ngseasy">NGSeasy</a> Docker <a href="https://registry.hub.docker.com/u/compbio/ngseasy-bwa/">compbio/ngseasy-bwa</a> image.</p>

<blockquote>
<p>Program: bwa (alignment via Burrows-Wheeler transformation)<br>
Version: 0.7.12-r1039<br>
Contact: Heng Li <a href="mailto:lh3@sanger.ac.uk">lh3@sanger.ac.uk</a>  </p>
</blockquote>

<div class="highlight highlight-bash"><pre><span class="pl-c">###################################</span>
<span class="pl-c">## BWA &gt; samblaster &gt; samtools</span>
<span class="pl-c">#</span>

array=<span class="pl-s"><span class="pl-pds">"</span><span class="pl-pds">"</span></span>
refGenome=<span class="pl-s"><span class="pl-pds">"</span><span class="pl-pds">"</span></span>

docker run \
-w /home/pipeman \
-e HOME=/home/pipeman \
-e USER=pipeman \
--user pipeman \
-i \
-t compbio/ngseasy-bwa:1.0 /bin/bash -c \
<span class="pl-s"><span class="pl-pds">"</span>bwa mem -t 32 -V -M -a <span class="pl-smi">${refGenome}</span> <span class="pl-smi">${array}</span>.fasta | <span class="pl-cce">\</span></span>
<span class="pl-s">samblaster --addMateTags --excludeDups | <span class="pl-cce">\</span></span>
<span class="pl-s">samtools sort -@ 32 -T temp_ -O sam -o <span class="pl-smi">${array}</span>.sam &amp;&amp; <span class="pl-cce">\</span></span>
<span class="pl-s">samtools index <span class="pl-smi">${array}</span>.sam &amp;&amp; <span class="pl-cce">\</span></span>
<span class="pl-s">rm <span class="pl-smi">${array}</span>.sam<span class="pl-pds">"</span></span></pre></div>

<p><strong>MEGA</strong>    </p>

<ul>
<li>
<a href="https://s3-eu-west-1.amazonaws.com/illumina-probe-mappings/mega_array_seq_sorted.bam">mega_array_seq_sorted.bam</a><br>
</li>
<li>
<a href="https://s3-eu-west-1.amazonaws.com/illumina-probe-mappings/mega_array_seq_sorted.bam.bai">mega_array_seq_sorted.bam.bai</a><br>
</li>
</ul>

<p><strong>OMNI</strong>  </p>

<ul>
<li>...<br>
</li>
<li>...<br>
</li>
</ul>

<h3>
<a id="probe-lists" class="anchor" href="#probe-lists" aria-hidden="true"><span class="octicon octicon-link"></span></a>Probe Lists</h3>

<p>These lists provide data on probe mappings. We provide Illumina Probe Id's along with the
number of time it maps to the genome. </p>

<table>
<thead>
<tr>
<th>BeadArry</th>
<th>probe mapping</th>
<th>good probes</th>
</tr>
</thead>
<tbody>
<tr>
<td>MEGA_Consortium_15063755_B2</td>
<td><a href="">xxx</a></td>
<td><a href="">xxx</a></td>
</tr>
<tr>
<td>HumanOmniExpressExome-8 v1.1</td>
<td><a href="">xxx</a></td>
<td><a href="">xxx</a></td>
</tr>
<tr>
<td>HumanOmniExpress-24 v1.1</td>
<td><a href="">xxx</a></td>
<td><a href="">xxx</a></td>
</tr>
<tr>
<td>HumanOmniExpress-24 v1.0</td>
<td><a href="">xxx</a></td>
<td><a href="">xxx</a></td>
</tr>
<tr>
<td>HumanCoreExome-24 v1.0</td>
<td><a href="">xxx</a></td>
<td><a href="">xxx</a></td>
</tr>
</tbody>
</table>

<h3>
<a id="table-1-number-of-variants-after-probe-mapping" class="anchor" href="#table-1-number-of-variants-after-probe-mapping" aria-hidden="true"><span class="octicon octicon-link"></span></a>Table 1. Number of variants after probe mapping</h3>

<table>
<thead>
<tr>
<th>BeadArry</th>
<th>N SNP Start</th>
<th>N SNP End</th>
</tr>
</thead>
<tbody>
<tr>
<td>MEGA_Consortium_15063755_B2</td>
<td>1.7m</td>
<td>1.5m</td>
</tr>
<tr>
<td>HumanOmniExpressExome-8 v1.1</td>
<td>?</td>
<td>?</td>
</tr>
<tr>
<td>HumanOmniExpress-24 v1.1</td>
<td>?</td>
<td>?</td>
</tr>
<tr>
<td>HumanOmniExpress-24 v1.0</td>
<td>?</td>
<td>?</td>
</tr>
<tr>
<td>HumanCoreExome-24 v1.0</td>
<td>?</td>
<td>?</td>
</tr>
</tbody>
</table>

<h2>
<a id="2-genomestudio" class="anchor" href="#2-genomestudio" aria-hidden="true"><span class="octicon octicon-link"></span></a>2. Genomestudio</h2>

<ul>
<li>Manual clustering, inspection and filtering of variants in genomestudio</li>
<li>ensures that the most robust data is produced</li>
<li>allows iterative qc of samples and snps and ability to rescue data close to qc thresholds for exclusion</li>
<li>for each array we cluster variants based on data and generate new custom egt files </li>
<li>custom egt files used for all subsequent projects</li>
</ul>

<p>More details soon....</p>

<h2>
<a id="3-quality-control-and-re-calling" class="anchor" href="#3-quality-control-and-re-calling" aria-hidden="true"><span class="octicon octicon-link"></span></a>3. Quality Control and Re-Calling</h2>

<ul>
<li>data exported and processed using <a href="">custom scripts</a>
</li>
<li>standard gwas qc 

<ul>
<li>sample &amp; snp call rates</li>
<li>IBD </li>
<li>het if requested</li>
<li>maf &amp; hwe if requested</li>
<li>gender checks if requested and if gender provided</li>
<li>No PCA or MDS (this is for the data owners to do)</li>
</ul>
</li>
<li>No-Call Variants recalled using zCall</li>
<li>Produce PLINK Files for further analysis by data owners </li>
</ul>

<p>More details soon....</p>

<hr>

<h3>
<a id="no-genotypes-and-samples-processed" class="anchor" href="#no-genotypes-and-samples-processed" aria-hidden="true"><span class="octicon octicon-link"></span></a>No. Genotypes and Samples Processed</h3>

<table>
<thead>
<tr>
<th>Samples/SNPs</th>
<th>Total</th>
</tr>
</thead>
<tbody>
<tr>
<td>Samples</td>
<td>10000</td>
</tr>
<tr>
<td>SNPS</td>
<td>1000000000</td>
</tr>
</tbody>
</table>

<hr>

<p>Copyright (C) 2015 Hamel Patel, Amos Folarin &amp; Stephen Jeffrey Newhouse</p>

      <footer class="site-footer">
        <span class="site-footer-owner"><a href="https://github.com/KHP-Informatics/illumina-array-protocols">Illumina-array-protocols</a> is maintained by <a href="https://github.com/KHP-Informatics">KHP-Informatics</a>.</span>

        <span class="site-footer-credits">This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the <a href="https://github.com/jasonlong/cayman-theme">Cayman theme</a> by <a href="https://twitter.com/jasonlong">Jason Long</a>.</span>
      </footer>

    </section>

  
  </body>
</html>

